### PR TITLE
nanocoap sock: stop retransmissions after empty ACK

### PR DESCRIPTION
**Pull request by fabian18**
### Contribution description

The `nanocoap_sock_request_cb()` function does a synchronous CoAP request.
This PR fixes that further retransmissions are done, even though a matching empty ACK was received.
The behavior is mandatory for CoAP and should also be done when no response callback is set, because the ACK is not the actual response.

### Testing procedure

Run `examples/nanocoap_server` and `tests/net/nanocoap_cli` and increase the timeout for the `/separate` resource.
Without this fix, the client keeps retransmitting even after the empty ACK.

 Test when I set the server separate response delay to 10 seconds.

```
client get fe80::4d3:11ff:fe35:6a43 5683 /well-known/core
client get fe80::4d3:11ff:fe35:6a43 5683 /well-known/core
nanocli: sending msg ID 1, 22 bytes
nanocli: msg send failed: -6
> client get fe80::4d3:11ff:fe35:6a43 5683 /.well-known/core
client get fe80::4d3:11ff:fe35:6a43 5683 /.well-known/core
nanocli: sending msg ID 1, 23 bytes
nanocli: response Success, code 2.05, 16 bytes
</vfs>,</separat
> client get fe80::4d3:11ff:fe35:6a43 5683 /separate
client get fe80::4d3:11ff:fe35:6a43 5683 /separate
nanocli: sending msg ID 1, 15 bytes
nanocli: response Success, code 2.05, 28 bytes
00000000  54  68  69  73  20  69  73  20  61  20  64  65  6C  61  79  65
00000010  64  20  72  65  73  70  6F  6E  73  65  2E  00
```
~~bug: Look at `</vfs>,</separat`~~ no bug just blockwise transfer

When I disable the ACK on the server side I can see retransmissions in wireshark:

Screenshot from 2024-05-27 11-46-07

But another bug: The serveice unavailable ACK is treated as the response. :( 

### Issues/PRs references

first split out of #20696

### CODE REVIEW

**Comment by fabian18** [comment]
need to think if the state handling is done correct here.`state = STATE_RESPONSE_RCVD;` for everything that is received

File: sys/net/application_layer/nanocoap/sock.c [hunk 1]
```diff
@@ -267,7 +267,6 @@ ssize_t nanocoap_sock_request_cb(nanocoap_sock_t *sock, coap_pkt_t *pkt,
                 DEBUG("nanocoap: timeout waiting for response\n");
                 timeout *= 2;
                 deadline = _deadline_from_interval(timeout);
-                state = STATE_REQUEST_SEND;
```
Comment (benpicco) [inline sys/net/application_layer/nanocoap/sock.c:270]: Won't this disable *all* retransmissions now?

**Comment by benpicco** [comment]
> Can I fix it right here in this PR?

Sure!

File: sys/net/application_layer/nanocoap/sock.c [hunk 2]
```diff
@@ -267,7 +280,6 @@ ssize_t nanocoap_sock_request_cb(nanocoap_sock_t *sock, coap_pkt_t *pkt,
...
                 continue;
```
Comment (benpicco) [inline sys/net/application_layer/nanocoap/sock.c:283]: We are now stuck in this loop if there really is no response from the server. (e.g. we send a request to a non-existing address)

Comment (benpicco) [inline sys/net/application_layer/nanocoap/sock.c:283]: If the separate response never arrives, we are forever stuck in this loop.

Comment (benpicco) [reply sys/net/application_layer/nanocoap/sock.c:283]:
```patch
--- a/sys/net/application_layer/nanocoap/sock.c
+++ b/sys/net/application_layer/nanocoap/sock.c
@@ -225,10 +225,6 @@ ssize_t nanocoap_sock_request_cb(nanocoap_sock_t *sock, coap_pkt_t *pkt,
     while (1) {
         switch (state) {
         case STATE_REQUEST_SEND:
-            if (tries_left == 0) {
-                DEBUG("nanocoap: maximum retries reached\n");
-                return -ETIMEDOUT;
-            }
             --tries_left;
 
             DEBUG("nanocoap: send %u bytes (%u tries left)\n",
@@ -277,6 +273,13 @@ ssize_t nanocoap_sock_request_cb(nanocoap_sock_t *sock, coap_pkt_t *pkt,
             }
             res = tmp;
             if (res == -ETIMEDOUT) {
+                if (tries_left == 0) {
+                    DEBUG("nanocoap: maximum retries reached\n");
+                    state = STATE_STOP_RETRANSMIT;
+                    res = -ETIMEDOUT;
+                    continue;
+                }
+
                 DEBUG("nanocoap: timeout waiting for response\n");
                 timeout *= 2;
                 deadline = _deadline_from_interval(timeout);
```

should fix this

Comment (fabian18) [reply sys/net/application_layer/nanocoap/sock.c:284]: yup thanks.
I just tested it. Also when no ACK is sent the retransmissions will time out.

**Comment by fabian18** [comment]
Given that the request contains a no response option to suppresses certain response classes, this must be communicated to the function somehow. Even if there is no callback, maybe the program must wait for a response.

I think before the assumption was, if the no-response option is included, the caller does not provide a callback.

**Review by benpicco — APPROVED** [review]
So you suggest we should differentiate between

 - ACK received but no separate response (because it was suppressed or lost)  
 - no ACK received

But I think this is a separate issue.

**Comment by fabian18** [comment]
Thanks for approval.

Luckily we only add a no-response option for NON without buffer to write the response to. This case is still catched.
> 
    - ACK received but no separate response (because it was suppressed or lost)
    - no ACK received

That I did not event think of. But yes, if there is a timeout it could be due to no ACK or no response, which means different things.

> But I think this is a separate issue.

Yes

**Pull request merged**